File: php/nhlapidocumentation_sdk.php
<?php
declare(strict_types=1);

// NhlApiDocumentation SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class NhlApiDocumentationSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new NhlApiDocumentationUtility();
        $this->_utility = $utility;

        $config = NhlApiDocumentationConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = NhlApiDocumentationHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = NhlApiDocumentationHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, NhlApiDocumentationFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return NhlApiDocumentationUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = NhlApiDocumentationHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = NhlApiDocumentationHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = NhlApiDocumentationHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new NhlApiDocumentationSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = NhlApiDocumentationHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = NhlApiDocumentationHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_conference = null;

    // Idiomatic facade: $client->conference()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Conference() (PHP method
    // names are case-insensitive).
    public function conference($data = null)
    {
        require_once __DIR__ . '/entity/conference_entity.php';
        if ($data === null) {
            if ($this->_conference === null) {
                $this->_conference = new ConferenceEntity($this, null);
            }
            return $this->_conference;
        }
        return new ConferenceEntity($this, $data);
    }


    private $_division = null;

    // Idiomatic facade: $client->division()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Division() (PHP method
    // names are case-insensitive).
    public function division($data = null)
    {
        require_once __DIR__ . '/entity/division_entity.php';
        if ($data === null) {
            if ($this->_division === null) {
                $this->_division = new DivisionEntity($this, null);
            }
            return $this->_division;
        }
        return new DivisionEntity($this, $data);
    }


    private $_game = null;

    // Idiomatic facade: $client->game()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Game() (PHP method
    // names are case-insensitive).
    public function game($data = null)
    {
        require_once __DIR__ . '/entity/game_entity.php';
        if ($data === null) {
            if ($this->_game === null) {
                $this->_game = new GameEntity($this, null);
            }
            return $this->_game;
        }
        return new GameEntity($this, $data);
    }


    private $_player = null;

    // Idiomatic facade: $client->player()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Player() (PHP method
    // names are case-insensitive).
    public function player($data = null)
    {
        require_once __DIR__ . '/entity/player_entity.php';
        if ($data === null) {
            if ($this->_player === null) {
                $this->_player = new PlayerEntity($this, null);
            }
            return $this->_player;
        }
        return new PlayerEntity($this, $data);
    }


    private $_player_stat = null;

    // Idiomatic facade: $client->player_stat()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias PlayerStat() (PHP method
    // names are case-insensitive).
    public function player_stat($data = null)
    {
        require_once __DIR__ . '/entity/player_stat_entity.php';
        if ($data === null) {
            if ($this->_player_stat === null) {
                $this->_player_stat = new PlayerStatEntity($this, null);
            }
            return $this->_player_stat;
        }
        return new PlayerStatEntity($this, $data);
    }


    private $_roster = null;

    // Idiomatic facade: $client->roster()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Roster() (PHP method
    // names are case-insensitive).
    public function roster($data = null)
    {
        require_once __DIR__ . '/entity/roster_entity.php';
        if ($data === null) {
            if ($this->_roster === null) {
                $this->_roster = new RosterEntity($this, null);
            }
            return $this->_roster;
        }
        return new RosterEntity($this, $data);
    }


    private $_schedule = null;

    // Idiomatic facade: $client->schedule()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Schedule() (PHP method
    // names are case-insensitive).
    public function schedule($data = null)
    {
        require_once __DIR__ . '/entity/schedule_entity.php';
        if ($data === null) {
            if ($this->_schedule === null) {
                $this->_schedule = new ScheduleEntity($this, null);
            }
            return $this->_schedule;
        }
        return new ScheduleEntity($this, $data);
    }


    private $_standing = null;

    // Idiomatic facade: $client->standing()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Standing() (PHP method
    // names are case-insensitive).
    public function standing($data = null)
    {
        require_once __DIR__ . '/entity/standing_entity.php';
        if ($data === null) {
            if ($this->_standing === null) {
                $this->_standing = new StandingEntity($this, null);
            }
            return $this->_standing;
        }
        return new StandingEntity($this, $data);
    }


    private $_team = null;

    // Idiomatic facade: $client->team()->list() / ->load(["id" => ...]).
    // Also serves the deprecated PascalCase alias Team() (PHP method
    // names are case-insensitive).
    public function team($data = null)
    {
        require_once __DIR__ . '/entity/team_entity.php';
        if ($data === null) {
            if ($this->_team === null) {
                $this->_team = new TeamEntity($this, null);
            }
            return $this->_team;
        }
        return new TeamEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new NhlApiDocumentationSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}

File: rb/NhlApiDocumentation_sdk.rb
# NhlApiDocumentation SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'NhlApiDocumentation_types'


class NhlApiDocumentationSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = NhlApiDocumentationUtility.new
    @_utility = utility

    config = NhlApiDocumentationConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = NhlApiDocumentationHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = NhlApiDocumentationHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, NhlApiDocumentationFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    NhlApiDocumentationUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = NhlApiDocumentationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = NhlApiDocumentationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = NhlApiDocumentationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = NhlApiDocumentationSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue NhlApiDocumentationError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = NhlApiDocumentationHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = NhlApiDocumentationHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.Conference.list / client.Conference.load({ "id" => ... })
  def Conference(data = nil)
    require_relative 'entity/conference_entity'
    ConferenceEntity.new(self, data)
  end


  # Canonical facade: client.Division.list / client.Division.load({ "id" => ... })
  def Division(data = nil)
    require_relative 'entity/division_entity'
    DivisionEntity.new(self, data)
  end


  # Canonical facade: client.Game.list / client.Game.load({ "id" => ... })
  def Game(data = nil)
    require_relative 'entity/game_entity'
    GameEntity.new(self, data)
  end


  # Canonical facade: client.Player.list / client.Player.load({ "id" => ... })
  def Player(data = nil)
    require_relative 'entity/player_entity'
    PlayerEntity.new(self, data)
  end


  # Canonical facade: client.PlayerStat.list / client.PlayerStat.load({ "id" => ... })
  def PlayerStat(data = nil)
    require_relative 'entity/player_stat_entity'
    PlayerStatEntity.new(self, data)
  end


  # Canonical facade: client.Roster.list / client.Roster.load({ "id" => ... })
  def Roster(data = nil)
    require_relative 'entity/roster_entity'
    RosterEntity.new(self, data)
  end


  # Canonical facade: client.Schedule.list / client.Schedule.load({ "id" => ... })
  def Schedule(data = nil)
    require_relative 'entity/schedule_entity'
    ScheduleEntity.new(self, data)
  end


  # Canonical facade: client.Standing.list / client.Standing.load({ "id" => ... })
  def Standing(data = nil)
    require_relative 'entity/standing_entity'
    StandingEntity.new(self, data)
  end


  # Canonical facade: client.Team.list / client.Team.load({ "id" => ... })
  def Team(data = nil)
    require_relative 'entity/team_entity'
    TeamEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = NhlApiDocumentationSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
